### PR TITLE
Fix android device-related commands

### PR DESCRIPTION
**Pull request by teobugslayer**
Generate proper command line for strating adb
Change public API for adb-related functions
Use ES6 imports
When there's no content in http body , print [no content] instead of [nothing]

Fixes http://teampulse.telerik.com/view#item/299038

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:heart:

File: definitions/mobile.d.ts [hunk 1]
```diff
@@ -85,8 +85,8 @@ declare module Mobile {
 	}
 
 	interface IAndroidDebugBridge {
-		executeCommand(...args: string[]): IFuture<any>;
-		executeShellCommand(...args: string[]): IFuture<any>;
+		executeCommand(args: string[]): IFuture<any>;
+		executeShellCommand(args: string[]): IFuture<any>;
```
Comment (Fatme) [inline definitions/mobile.d.ts]: Why do you need to change this?

Comment (teobugslayer) [inline definitions/mobile.d.ts]: If i leave it as-was, the compiler complains with `error TS2345: Argument of type 'string[]' is not assignable to parameter of type 'string'.` I am open to suggestions how else to fix it.

File: mobile/android/android-application-manager.ts [hunk 2]
```diff
@@ -25,27 +25,27 @@ export class AndroidApplicationManager implements Mobile.IDeviceApplicationManag
 			return this._installedApplications;
 		}).future<string[]>()();
 	}
-	
+
 	public installApplication(packageFilePath: string): IFuture<void> {
-		return this.adb.executeCommand(`install -r "${packageFilePath}"`);
+		return this.adb.executeCommand(["install", "-r", `${packageFilePath}`]);
```
Comment (rosen-vladimirov) [inline mobile/android/android-application-manager.ts]: if packageFilePath contains spaces, we may still fail, maybe we should verify this and use something like:
`"${packageFilePath}"`

File: mobile/android/android-debug-bridge.ts [hunk 3]
```diff
@@ -11,35 +11,34 @@ interface IComposeCommandResult {
...
+			return this.$childProcess.spawnFromEvent(shellCommand.command, shellCommand.args, "close", undefined, {throwError: false}).wait().stdout;
 		}).future<any>()();
 	}
-	
+
 	public sendBroadcastToDevice(action: string, extras: IStringDictionary = {}): IFuture<number> {
 		return (() => {
-			let broadcastCommand = `am broadcast -a "${action}"`;
-			_.each(extras, (value,key) => broadcastCommand += ` -e "${key}" "${value}"`);
+			let broadcastCommand = ["am", "broadcast", "-a", `${action}`];
```
Comment (rosen-vladimirov) [inline mobile/android/android-debug-bridge.ts]: is there any chance to have space in "action", as far as I see from the previous code, it was intended to have "", so consider using `"${action}"`

File: mobile/android/android-device-file-system.ts [hunk 4]
```diff
@@ -20,17 +20,17 @@ export class AndroidDeviceFileSystem implements Mobile.IDeviceFileSystem {
...
 		return (() => { }).future<void>()();
 	}
-	
+
 	public transferFiles(appIdentifier: string, localToDevicePaths: Mobile.ILocalToDevicePathData[]): IFuture<void> {
 		return (() => {
 			_(localToDevicePaths)
 				.filter(localToDevicePathData => this.$fs.getFsStats(localToDevicePathData.getLocalPath()).wait().isFile())
-				.each(localToDevicePathData => this.adb.executeCommand(`push "${localToDevicePathData.getLocalPath()}" "${localToDevicePathData.getDevicePath()}"`).wait())
+				.each(localToDevicePathData => this.adb.executeCommand(["push", `${localToDevicePathData.getLocalPath()}`, `${localToDevicePathData.getDevicePath()}`]).wait())
```
Comment (rosen-vladimirov) [inline mobile/android/android-device-file-system.ts]: we have to add "" around `localToDevicePathData.getLocalPath()` and `localToDevicePathData.getDevicePath()`

File: mobile/android/android-device-file-system.ts [hunk 5]
```diff
@@ -20,17 +20,17 @@ export class AndroidDeviceFileSystem implements Mobile.IDeviceFileSystem {
...
 				.value();
-			
+
 			_(localToDevicePaths)
 				.filter(localToDevicePathData => this.$fs.getFsStats(localToDevicePathData.getLocalPath()).wait().isDirectory())
-				.each(localToDevicePathData => this.adb.executeShellCommand(`chmod 0777 "${localToDevicePathData.getDevicePath()}"`).wait())
+				.each(localToDevicePathData => this.adb.executeShellCommand(["chmod", "0777", `${localToDevicePathData.getDevicePath()}`]).wait())
```
Comment (rosen-vladimirov) [inline mobile/android/android-device-file-system.ts]: same as above - "" around `localToDevicePathData.getDevicePath()`

File: mobile/android/android-device-file-system.ts [hunk 6]
```diff
@@ -40,10 +40,10 @@ export class AndroidDeviceFileSystem implements Mobile.IDeviceFileSystem {
 			this.$logger.trace(`Transfering ${localPath} to ${devicePath}`);
 			let stats = this.$fs.getFsStats(localPath).wait();
 			if(stats.isDirectory()) {
-				this.adb.executeShellCommand(`mkdir "${path.dirname(devicePath)}"`).wait();
+				this.adb.executeShellCommand(["mkdir", `${path.dirname(devicePath)}`]).wait();
```
Comment (rosen-vladimirov) [inline mobile/android/android-device-file-system.ts]: still not sure about this, but maybe we need "" as well

File: definitions/mobile.d.ts (hunk 1 above)
Comment (ligaz) [inline definitions/mobile.d.ts]: Are you both using the same version of the TypeScript compiler?

Comment (Fatme) [inline definitions/mobile.d.ts]: Yes, we use the same version of the TypeScript compiler. The problem actually comes after this change https://github.com/telerik/mobile-cli-lib/commit/aa5e00d3e8c0113837aa8828a746f3d0ea2446bb#diff-416bc024035578187882887a05fe7bfaR22. We've replaced `exec` with `spawnFromEvent`.

File: mobile/android/android-debug-bridge.ts [hunk 7]
```diff
@@ -11,35 +11,34 @@ interface IComposeCommandResult {
...
+			_.each(extras, (value,key) => broadcastCommand.concat(["-e", `${key}`, `${value}`]));
```
Comment (rosen-vladimirov) [inline mobile/android/android-debug-bridge.ts]: is it possible to have spaces in key or value?

File: mobile/android/android-device-file-system.ts [hunk 8]
```diff
@@ -40,10 +40,10 @@ export class AndroidDeviceFileSystem implements Mobile.IDeviceFileSystem {
...
 			} else {
-				this.adb.executeCommand(`push "${localPath}" "${devicePath}"`).wait(); 
+				this.adb.executeCommand(["push", `${localPath}`, `${devicePath}`]).wait();
```
Comment (rosen-vladimirov) [inline mobile/android/android-device-file-system.ts]: and same here

File: mobile/android/android-livesync-service.ts [hunk 9]
```diff
@@ -25,39 +24,39 @@ class LiveSyncCommands {
...
 		return (() => {
 			let hostTmpDir = this.getTempDir();
 			let commandsFileHostPath = path.join(hostTmpDir, AndroidLiveSyncService.COMMANDS_FILE);
 			this.$fs.writeFile(commandsFileHostPath, commands.join("\n")).wait();
-			
+
 			// copy it to the device
 			this.device.fileSystem.transferFile(commandsFileHostPath, commandsFileDevicePath).wait();
-			this.device.adb.executeShellCommand(`chmod 0777 ${commandsFileDevicePath}`).wait();
+			this.device.adb.executeShellCommand(["chmod", "0777", `${commandsFileDevicePath}`]).wait();
```
Comment (rosen-vladimirov) [inline mobile/android/android-livesync-service.ts]: maybe "${commandsFileDevicePath}"

File: mobile/android/android-debug-bridge.ts (hunk 7 above)
Comment (teobugslayer) [inline mobile/android/android-debug-bridge.ts]: No problem, spawn will add double quotes.

**Comment by rosen-vladimirov** [comment]
Ok, after discussion, we have decided that my comments are incorrect as spawn guarantees that an argument, which has spaces in it, will be sent correctly to the command. So you have my :+1:

**Comment by teobugslayer** [comment]
:+1: from @fatme as well.

**Pull request merged**